Consuming APIs in flutter (#1862)
section-engineering-education/engineering-education: articles/consuming-apis-in-flutter/index.md

Application Programming Interface (API) is a communication portal that allows two or more applications to connect for data sharing. It acts as an intermediary for delivering requests to service providers and returning the responses. The use of APIs has gained prevalence in mobile application development, given the ease of using pre-existing frameworks. Programmers use most APIs to fetch data from web servers and render it to the application's UI components.

> This article will demonstrate how to fetch and consume data from a server using a RESTful API in a flutter application. You can download the code for this application form [here](https://github.com/jerimkaura/flutter-book/tree/main/mygithub).

Reviewer: The GitHub link specified here returns a 404.

![image](https://user-images.githubusercontent.com/34317373/112730676-e4981c00-8f58-11eb-8604-28b192197089.png)

> This article will demonstrate how to fetch and consume data from a server using a RESTful API in a flutter application. You can download the code for this application form [here](https://github.com/jerimkaura/flutter-book/tree/main/mygithub).

Reviewer: ```suggestion
This article will demonstrate how to fetch and consume data from a server using a RESTful API in a flutter application. You can download the code for this application from [here](https://github.com/jerimkaura/flutter-book/tree/main/mygithub).
```

> This article will demonstrate how to fetch and consume data from a server using a RESTful API in a flutter application. You can download the code for this application form [here](https://github.com/jerimkaura/flutter-book/tree/main/mygithub).

Reviewer: Please explain briefly what a RESTful API is, it would add more value to the article

### Prerequisites
1. A basic understanding of Flutter

> 3. Flutter SDK installed on your computer

Reviewer: Please make sure the order of the numbered list is correct. You've entered it here as 1, 3, 2, 3

> 3. Flutter SDK installed on your computer

Reviewer: It might be helpful to include a link that describes how to set up the Flutter SDK on your computer

2. Code editor, [Android Studio](https://developer.android.com/studio), or [VSCode](https://code.visualstudio.com/download) are most preferred.
3. An emulator or a mobile device to run the code.

### Table oF contents
- Setting up the application
- The HTTP package
- Creating Data classes from JSON
- Implementing the providers
- Consuming the data returned from the API request on a mobile screen
- Conclusion
- Further Reading

> 

Reviewer: Please include internal links to each heading in the table of contents

Example,

```markdown
- [Setting up the application](#setting-up-the-application)
```

### Setting up the application

> We will build our application based on the Github API. Therefore, we need to obtain the GitHub `client key` and `secret` to access the API. Check [this link](https://www.knowband.com/blog/user-manual/get-github-client-id-client-secret-api-details/) on a complete guide on getting the `client key` and `secret`.

Reviewer: ```suggestion
We will build our application based on the Github API. Therefore, we need to obtain the GitHub `client key` and `secret` to access the API. Check [this link](https://www.knowband.com/blog/user-manual/get-github-client-id-client-secret-api-details/) for a complete guide on getting the `client key` and `secret`.
```

We need to organize the folders within our flutter project to locate our application's files and components with ease. This practice makes us find bugs easily, unlike writing the entire application on a single file. You can check [this](https://www.section.io/engineering-education/flutter-folder-organization/) article for the preferable folder organization. The final folder organization should appear as below:

> ```

Reviewer: How does this folder structure arise? Please explain how to set up the Flutter project from scratch, since this article is targeted towards beginners

lib
    ┣ models
    ┃ ┗ User.dart
    ┣ Providers
    ┃ ┗ UserProvider.dart
    ┣ Requests
    ┃ ┗ GithubRequest.dart
    ┣ Screens
    ┃ ┗ FollowersPage.dart
    ┗ main.dart
```

### Adding the HTTP Package
The `HTTP package` contains a set of high-level functions for use in HTTP resource consumption. To add the package to our application, open `pubspec.yml` and add the following line under `dependencies:

```yml
dependencies:
flutter:
    SDK: flutter
HTTP: ^0.12.2
```

Next, we will import the HTTP package into our `GithubRequest.dart` file with the following line of code:

```dart
import 'package:HTTP/HTTP.dart' as HTTP;
```

The below shows how we will use the package to fetch data from the API.

> 

Reviewer: Please mention that the below code must be under the `GitHubRequest.dart` file. A beginner may get confused

Since flutter accepts dart as the primary programming language, we need to convert the JSON data fetched from the URL to dart Classes for consumption in the application. We can do that using the [Quicktype](https://app.quicktype.io/) website where we pass the JSON object, and a class of the object is returned based on a specified language. Will will return our classes in `dart`. For instance, our JSON representing the User is as shown below:

```dart
{
  "login": "jerimkaura",
  "avatar_url": "https://avatars.githubusercontent.com/u/50904889?v=4",
  "location": "Nairobi"
}
```

#### The user class
I edited the JSON to capture only the attributes needed on the application. When we pass the above JSON into [Quicktype](https://app.quicktype.io/), the generated user class is as below:

```dart
// To parse this JSON data, do
//     final user = userFromJson(jsonString);

import 'dart:convert';

//the created user class
class User {
  //class constructor
  User({
    this.login,
    this.avatarUrl,
    this.location,
  });

  String login;
  String avatarUrl;
  String location;

  //JSON serialization
  factory User.fromRawJson(String str) => User.fromJson(json.decode(str));

>   String toRawJson() => json.encode(toJson());

Reviewer: Please add more code comments in this snippet

  factory User.fromJson(Map<String, dynamic> json) => User(
    login: json["login"],
    avatarUrl: json["avatar_url"],
    location: json["location"],
  );

  Map<String, dynamic> toJson() => {
    "login": login,
    "avatar_url": avatarUrl,
    "location": location,
  };
}
```

### Adding the providers
The Provider will have the functions required to fetch the API's user data and deliver a response. We will create a file called `UserProvider.dart` under the `Providers` folder. The `ChangeNotifier` class will notify our view when one more variable changes.

```dart

class UserProvider with ChangeNotifier {
  User user; //an instance of a user
  String errorMessage; //error message
  bool loading = false; //loading the page

>   Future<bool> fetchUser(username) async {

Reviewer: it might be a good idea to explain why we are using an Async method here.
Also, please include more code comments that effectively explains all the lines

    return loading; //return true if the app is loading the data
  }

  void setLoading(value) {
    loading = value;
    notifyListeners();
  }

  void setUser(value) {
    user = value;
    notifyListeners();
  }

  User getUser() {
    return user; //returns the fetched user
  }

  void setMessage(value) {
    errorMessage = value;
    notifyListeners();
  }

  String getMessage() {
    return errorMessage;
  }

  bool isUser() {
    return user != null ? true : false; // returns true if user is not null, anf false otherwise
  }
}

``` 

### Consuming the data

With our model and Provider ready, our application will fetch data via the Provider and convert the JSON result into dart classes using the User model's methods. The next thing we are doing is consuming the data on a mobile screen. For this process, we will do three primary things.

#### Instantiate the user classes.
We will have two instances of the user class; one instance is for the User and another as a list of followers of a given user.

```dart
User user; //instantiate a user
List< User> followers; // instantiate a list of users as a placeholder for the followers.
```

#### Fetching data using setState()

> The setState method notifys the application that the internal state of the application has been changed and that the change might affetc the view.

Reviewer: ```suggestion
The setState method notifies the application that the internal state of the application has been changed and that the change might affect the view.
```

> ```dart

Reviewer: Please do mention in which file we have to include this code
